### DISCUSSION TITLE
Mathematica's native Manipulate and Animate

### DISCUSSION THREAD

**Discussion by gganapat** [WLJSTeam/wljs-notebook#389] [Graphics & Plot functions]

**Comment by gganapat** [comment]
ok, sorry. I updated to the latest and it looks good now! With MMAView and
this update, (and so many other changes), this is an absolutely impressive
piece of work!
Any chance, MMAView can be integrated into animation like Manipulate?

Thanks a bunch!
gani -

**Comment by JerryI** [comment]
> ok, sorry. I updated to the latest and it looks good now! With MMAView and
> this update, (and so many other changes), this is an absolutely impressive
> piece of work!
> Any chance, MMAView can be integrated into animation like Manipulate?
> 
> Thanks a bunch!
> gani -
> 
> On Wed, May 7, 2025 at 12:03 PM Gani Ganapathi ***@***.***>
> wrote:
> […](#)

Thanks! :)

> Any chance, MMAView can be integrated into animation like Manipulate?

this is a small problem with a async operations in Wolfram Kernel, since `MMAView` uses a parallel kernel to perform the evaluation and it messes up with update cycle of the sliders.

Anyway we will think about that.

However for the animations this must be possible using raster animation (sort of a live video)

```mathematica
frames = Table[MMAView @ Plot[Sin[y x], {y,0,10}], {x, 0.1, 1.0, 0.05}];

Module[{buffer = ImageData[frames//First, "Byte"]},
  Animate[Image[buffer // Offload, "Byte"], {i, 1, Length[frames], 1},
    "UpdateFunction" -> Function[i,
      (* stabilize raw image size *)
      buffer = ImageData[ImageCrop[frames[[i]], ImageDimensions[frames // First]], "Byte"];
      False (* prevent full refresh *)
    ]
  ]
]
```

**Comment by JerryI** [comment]
We should make somewhat like `MMAManipulate` for the full convince

**Comment by JerryI** [comment]
With a bit of a polishing it can become quite useful

*proof of concept*

```mathematica

With[{function = Plot[Sin[#1 x], {x,0,10}]&},
 With[{firstImage = MMAView[function[1]]},
  With[{dims = ImageDimensions[firstImage]},
    Module[{buffer = ImageData[firstImage, "Byte"]},
      {
        EventHandler[InputRange[0, 2, 0.1, "Label"->"w"], Function[w,
          Then[ParallelSubmitFunctionAsync[Function[{args, cbk},
              cbk @ ImageData[ImageCrop[Rasterize[function @ w], dims], "Byte"]
            ], {all}], Function[imgData,
              buffer = imgData;
          ]];
        ]],

        Image[buffer // Offload, "Byte"]
      } // Column 

    ]
  ]
 ]
]
```

https://github.com/user-attachments/assets/9c594c7d-052e-427a-aa10-8d66220e0ca2

**Comment by gganapat** [comment]
Nice! It works exactly like how I would have wanted it to be! The next
frontier - 3D plots? Is that even possible? Like viewing it from different
angles with mouse movement?

gani -

**Comment by JerryI** [comment]
Yep. One can overlay it with Graphics, then it is possible to attach event listeners to the mouse position and clicks. The rest is the same as for the previous example I gave.

I will try it tomorrow

**Comment by JerryI** [comment]
Looks ugly, but as proof of concept it works ;)


https://github.com/user-attachments/assets/a19667aa-54b2-43c7-9326-1cfb6e82a288



```mathematica
getAngles[angles_] := 3.0{Cos[-angles[[1]]]Sin[angles[[2]]], Sin[-angles[[1]]]Sin[angles[[2]]], Cos[angles[[2]]]}

With[{function = Function[{w, vec}, 
  Plot3D[Sin[w x] Cos[(2 - w) y], {x,0,10}, {y,0,10}, ViewPoint->vec]
]},
 With[{firstImage = MMAView[function[1, getAngles[{-Pi/4, Pi/3}]]]},
  With[{dims = ImageDimensions[firstImage]},
    Module[{
      buffer = ImageData[firstImage, "Byte"], 
      target = {0,0}, start = {0,0}, moving = False,
      framesSkip = 0, args = {1.0, getAngles[{-Pi/4, Pi/3}]}//N, recalc,
      angles = {-Pi/4, Pi/3}
    },

      recalc := With[{},
        Then[ParallelSubmitFunctionAsync[Function[{arguments, cbk},
              cbk @ ImageData[ImageCrop[Rasterize[function @@ arguments], dims, Padding->White], "Byte"]
            ], args], Function[imgData,
              buffer = imgData;
        ]];          
      ];
    
      {
        EventHandler[InputRange[0, 2, 0.1, "Label"->"w"], Function[w,
          args[[1]] = w;
          recalc;
        ]],

        EventHandler[Graphics[{
          Inset[Image[buffer // Offload, "Byte"], {0,0}]
        }, 
          ImageSizeRaw -> ImageDimensions[firstImage], 
          PlotRange->{{-1,1}, {-1,1}},
          "Controls"->False
        ], {
          "mousemove" -> Function[xy, 
            target = xy; framesSkip++; 
            If[Mod[framesSkip, 3] == 0 && moving,
              With[{dir = (target - start)/5.0},
                angles += dir;
                args[[2]] = getAngles[angles];
                
              ];
              recalc;
            ]
          ],
          "mousedown" -> Function[xy, start = xy; moving = True],
          "mouseup" -> Function[xy, moving = False]
        }]
      } // Column 

    ]
  ]
 ]
]
```

**Reply by gganapat** [reply]
Really neat! You've managed to combine features of Manipulate and dynamic viewing of 3D plots!
gani -

**Reply by JerryI** [reply]
> Really neat! You've managed to combine features of Manipulate and dynamic viewing of 3D plots!
> 
> gani -

We will make a wrapper for sure :)

**Comment by JerryI** [comment]
Added in a new release

https://wljs.io/frontend/Reference/GUI/MMAView

**Comment by gganapat** [comment]
Very nice! Keep up the great work!

gani -
